server/src/main/java/io/druid/server/coordinator/CachingCostBalancerStrategyFactory.java

```diff
@@ -0,0 +1,152 @@
+      throw new ISE("CachingCostBalancerStrategyFactory is already started or has been stopped");
+    }
+    if (!started.compareAndSet(false, true)) {
+      throw new ISE("CachingCostBalancerStrategyFactory is already started");
+    }
+    serverInventoryView.registerSegmentCallback(
+        executorRef.get(),
+        new ServerView.SegmentCallback()
+        {
+          @Override
+          public ServerView.CallbackAction segmentAdded(
+              DruidServerMetadata server, DataSegment segment
+          )
+          {
+            clusterCostCacheBuilder.addSegment(server.getName(), segment);
+            return ServerView.CallbackAction.CONTINUE;
+          }
+
+          @Override
+          public ServerView.CallbackAction segmentRemoved(
+              DruidServerMetadata server, DataSegment segment
+          )
+          {
+            clusterCostCacheBuilder.removeSegment(server.getName(), segment);
+            return ServerView.CallbackAction.CONTINUE;
+          }
+
+          @Override
+          public ServerView.CallbackAction segmentViewInitialized()
+          {
+            initialized = true;
+            return ServerView.CallbackAction.CONTINUE;
+          }
+        }
+    );
+
+    serverInventoryView.registerServerCallback(
+        executorRef.get(),
+        server -> {
+          clusterCostCacheBuilder.removeServer(server.getName());
+          return ServerView.CallbackAction.CONTINUE;
+        }
+    );
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    if (!started.compareAndSet(true, false)) {
+      throw new ISE("CachingCostBalancerStrategyFactory is not started");
+    }
+    executorRef.get().shutdownNow();
+  }
+
+  @Override
+  public BalancerStrategy createBalancerStrategy(final ListeningExecutorService exec)
+  {
+    if (!initialized) {
+      // fallback to CostBalancerStrategy if it's not yet possible to create valid cluster cost cache
+      return new CostBalancerStrategy(exec);
+    }
+    try {
+      CompletableFuture<CachingCostBalancerStrategy> future = CompletableFuture.supplyAsync(
+          () -> new CachingCostBalancerStrategy(clusterCostCacheBuilder.build(), exec),
+          executorRef.get()
+      );
+      try {
+        return future.get(1, TimeUnit.SECONDS);
+      }
+      catch (CancellationException e) {
+        log.warn("CachingCostBalancerStrategy creation has been cancelled, fallback to CostBalancerStrategy");
+      } catch (ExecutionException e) {
+        log.warn(e, "Failed to create CachingCostBalancerStrategy, fallback to CostBalancerStrategy");
+      } catch (TimeoutException e) {
+        log.error("CachingCostBalancerStrategy creation took more than 1 second! Fallback to CostBalancerStrategy");
+      } catch (InterruptedException e) {
+        log.warn("CachingCostBalancerStrategy creation has been interrupted, fallback to CostBalancerStrategy");
+        Thread.currentThread().interrupt();
```

Comment:
interrupt status is cleared after InterruptedException thrown, restoring it's state allows to correctly handle thread interruption (if needed) on a higher level